### PR TITLE
Fixup DamagedByTerrain.

### LINKED ISSUE #10091: SpreadDamageWarhead impacts spread use ambiguity
I started looking into warhead impacts because of [Comment](https://github.com/OpenRA/OpenRA/pull/9357#issuecomment-159034056).
It appears that nearly all uses of warhead impacts uses the impact at position, with only ```PoisonedByTiberium``` and ```DamagedWithoutFoundation``` applying directly to actors (as well as the pending ```AreaBeam```).

Long story short, in the cases that ```SpreadDamageWarhead``` is applied directly to actors, its ```Spread``` and ```Range``` is effectively ignored entirely.

The suggestion is that ```SpreadDamageWarhead``` should always use its ```Spread```, and that a new warhead ```DirectDamageWarhead``` should be defined for cases where no spread is needed.

Hopefully this will help avoid confusion where it appears that spread is not being applied.

Search done in project for uses of warhead impact uses.
```
...OpenRA.Game\GameRules\WeaponInfo.cs - (153, 15) : public void Impact(Target target, Actor firedBy, IEnumerable<int> damageModifiers)
...OpenRA.Mods.Cnc\Effects\IonCannon.cs - (47, 12) : weapon.Impact(target, firedBy.PlayerActor, Enumerable.Empty<int>());
...OpenRA.Mods.Cnc\Traits\PoisonedByTiberium.cs - (51, 20) : Info.WeaponInfo.Impact(Target.FromActor(self), self.World.WorldActor, Enumerable.Empty<int>());
...OpenRA.Mods.Common\Activities\Air\FallToEarth.cs - (41, 27) : info.ExplosionWeapon.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
...OpenRA.Mods.Common\Effects\AreaBeam.cs - (181, 18) : args.Weapon.Impact(Target.FromActor(a), args.SourceActor, adjustedModifiers);
...OpenRA.Mods.Common\Effects\Bullet.cs - (209, 16) : args.Weapon.Impact(Target.FromPos(pos), args.SourceActor, args.DamageModifiers);
...OpenRA.Mods.Common\Effects\GravityBomb.cs - (78, 17) : args.Weapon.Impact(Target.FromPos(pos), args.SourceActor, args.DamageModifiers);
...OpenRA.Mods.Common\Effects\LaserZap.cs - (81, 17) : args.Weapon.Impact(Target.FromPos(target), args.SourceActor, args.DamageModifiers);
...OpenRA.Mods.Common\Effects\Missile.cs - (814, 16) : args.Weapon.Impact(Target.FromPos(pos), args.SourceActor, args.DamageModifiers);
...OpenRA.Mods.Common\Effects\NukeLaunch.cs - (84, 11) : weapon.Impact(Target.FromPos(pos), firedBy.PlayerActor, Enumerable.Empty<int>());
...OpenRA.Mods.Common\Traits\Buildings\Bridge.cs - (343, 29) : info.DemolishWeaponInfo.Impact(Target.FromPos(self.CenterPosition), saboteur, Enumerable.Empty<int>());
...OpenRA.Mods.Common\Traits\Crates\ExplodeCrateAction.cs - (38, 11) : weapon.Impact(Target.FromPos(collector.CenterPosition), collector, Enumerable.Empty<int>());
...OpenRA.Mods.Common\Traits\Explodes.cs - (84, 11) : weapon.Impact(Target.FromPos(self.CenterPosition), e.Attacker, Enumerable.Empty<int>());
...OpenRA.Mods.D2k\Traits\Buildings\DamagedWithoutFoundation.cs - (74, 20) : info.WeaponInfo.Impact(Target.FromActor(self), self.World.WorldActor, Enumerable.Empty<int>());
...OpenRA.Mods.RA\Effects\TeslaZap.cs - (58, 17) : args.Weapon.Impact(Target.FromPos(pos), args.SourceActor, args.DamageModifiers);
...OpenRA.Mods.RA\Traits\MadTank.cs - (129, 32) : info.DetonationWeaponInfo.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
...OpenRA.Mods.RA\Traits\MadTank.cs - (89, 33) : info.ThumpDamageWeaponInfo.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28546397-spreaddamagewarhead-impacts-spread-use-ambiguity?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #11383: PoisonedByTiberium can crash ScriptTriggers.Damaged
When testing https://github.com/OpenRA/OpenRA/pull/11368 I got a crash after attacking the lone grenadier. He came after me, ran over Tiberium and that causes it to be damaged by a null actor location:

```
Fatal Lua Error: [string "BindingSupport.lua"]:30: Uncaught CLR exception at Lua->CLR boundary: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.Actor.get_Location () [0x00007] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Actor.cs:60 
  at OpenRA.Mods.Common.Scripting.GeneralProperties.get_Location () <0x41592890 + 0x0001b> in <filename unknown>:0 
  at (wrapper delegate-invoke) <Module>:invoke_callvirt_CPos_GeneralProperties (OpenRA.Mods.Common.Scripting.GeneralProperties)
  at System.Reflection.MonoProperty.GetterAdapterFrame[T,R] (System.Reflection.Getter`2 getter, System.Object obj) [0x00000] in /home/abuild/rpmbuild/BUILD/mono-4.2.3/mcs/class/corlib/System.Reflection/MonoProperty.cs:329 
  at System.Reflection.MonoProperty.GetValue (System.Object obj, System.Object[] index) [0x0008e] in /home/abuild/rpmbuild/BUILD/mono-4.2.3/mcs/class/corlib/System.Reflection/MonoProperty.cs:397 
  --- End of inner exception stack trace ---
  at System.Reflection.MonoProperty.GetValue (System.Object obj, System.Object[] index) [0x000a1] in /home/abuild/rpmbuild/BUILD/mono-4.2.3/mcs/class/corlib/System.Reflection/MonoProperty.cs:399 
  at OpenRA.Scripting.ScriptMemberWrapper.Get (Eluant.LuaRuntime runtime) [0x00041] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Scripting/ScriptMemberWrapper.cs:104 
  at OpenRA.Scripting.ScriptObjectWrapper.get_Item (Eluant.LuaRuntime runtime, Eluant.LuaValue keyValue) [0x0002a] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Scripting/ScriptObjectWrapper.cs:58 
  at OpenRA.Actor.get_Item (Eluant.LuaRuntime runtime, Eluant.LuaValue keyValue) [0x0000e] in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Actor.cs:393 
  at Eluant.LuaRuntime.<IndexCallback>m__D (IList`1 toDispose) <0x41590440 + 0x00132> in <filename unknown>:0 
  at Eluant.LuaRuntime.LuaToClrBoundary (IntPtr state, Eluant.LuaToClrBoundaryCallback callback) <0x4158fe00 + 0x00159> in <filename unknown>:0 
  at OpenRA.Scripting.ScriptContext.FatalError (System.String message) [0x00001] in <filename unknown>:0 
  at OpenRA.Mods.Common.Scripting.ScriptTriggers.Damaged (OpenRA.Actor self, OpenRA.Traits.AttackInfo e) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Traits.Health.InflictDamage (OpenRA.Actor self, OpenRA.Actor attacker, Int32 damage, IWarhead warhead, Boolean ignoreModifiers) [0x00000] in <filename unknown>:0 
  at OpenRA.Actor.InflictDamage (OpenRA.Actor attacker, Int32 damage, IWarhead warhead) [0x00027] in <filename unknown>:0 
  at OpenRA.Mods.Common.Warheads.DamageWarhead.DoImpact (OpenRA.Actor victim, OpenRA.Actor firedBy, IEnumerable`1 damageModifiers) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Warheads.DamageWarhead.DoImpact (Target target, OpenRA.Actor firedBy, IEnumerable`1 damageModifiers) [0x00000] in <filename unknown>:0 
  at OpenRA.GameRules.WeaponInfo+<Impact>c__AnonStorey1.<>m__0 () [0x00027] in <filename unknown>:0 
  at OpenRA.GameRules.WeaponInfo.Impact (Target target, OpenRA.Actor firedBy, IEnumerable`1 damageModifiers) [0x00095] in <filename unknown>:0 
  at OpenRA.Mods.Cnc.Traits.PoisonedByTiberium.Tick (OpenRA.Actor self) [0x00000] in <filename unknown>:0 
  at OpenRA.World.<Tick>m__3 (TraitPair`1 x) [0x0000e] in <filename unknown>:0 
  at OpenRA.WorldUtils.DoTimed[T] (IEnumerable`1 e, System.Action`1 a, System.String text) [0x00022] in <filename unknown>:0 
  at OpenRA.World.Tick () [0x00122] in <filename unknown>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x00226] in <filename unknown>:0 
  at OpenRA.Game.LogicTick () [0x00055] in <filename unknown>:0 
  at OpenRA.Game.Loop () [0x000d4] in <filename unknown>:0 
  at OpenRA.Game.Run () [0x00042] in <filename unknown>:0 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in <filename unknown>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in <filename unknown>:0 
```

### PR DESCRIPTION
**Pull request by GraionDilach**
Depends on #11513.

Resolves #10091 and probably fixes #11383.

Closes #11720 as well.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Please make this hit next release.

**Comment by GraionDilach** [comment]
Removed the #11633 dependency because I have a feeling I can't hit that one within time.

**Comment by GraionDilach** [comment]
Rebased and dropped the commit #11653 also dealt with.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 1]
```diff
@@ -9,17 +9,24 @@
...
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("This actor receives damage from the given weapon when on the specified terrain type.")]
-	class DamagedByTerrainInfo : UpgradableTraitInfo, IRulesetLoaded, Requires<HealthInfo>
+	class DamagedByTerrainInfo : UpgradableTraitInfo, Requires<HealthInfo>
 	{
-		[Desc("The weapon which is used to damage the actor.")]
-		[WeaponReference, FieldLoader.Require] public readonly string Weapon;
+		[Desc("The damage the actor suffers on these terrain.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: `these` -> `this`

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs (hunk 1 above)
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: I don't think there is any need to be repetitive here, the user knows this relates to the given terrain types.
`Amount of damage received per tick.` would be more helpful.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Changed it to `Amount of damage received per DamageDelay ticks.` then.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 2]
```diff
@@ -9,17 +9,24 @@
...
-	class DamagedByTerrainInfo : UpgradableTraitInfo, IRulesetLoaded, Requires<HealthInfo>
+	class DamagedByTerrainInfo : UpgradableTraitInfo, Requires<HealthInfo>
 	{
-		[Desc("The weapon which is used to damage the actor.")]
-		[WeaponReference, FieldLoader.Require] public readonly string Weapon;
+		[Desc("Amount of damage received per DamageDelay ticks.")]
+		[FieldLoader.Require] public readonly int Damage = 0;
+
+		[Desc("Delay between receiving damage.")]
+		public readonly int DamageDelay = 0;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: This is an interval, not a delay.  Please rename to `DamageInterval` for consistency.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Done.

**Comment by GraionDilach** [comment]
Is there anything which needs to be fixed here atm?

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 3]
```diff
@@ -70,7 +74,7 @@ public void AddedToWorld(Actor self)
 			// Actors start with maximum damage applied
 			var delta = health.HP - damageThreshold;
 			if (delta > 0)
-				health.InflictDamage(self, self.World.WorldActor, new Damage(delta), false);
+				self.InflictDamage(self, new Damage(delta));
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Hm, just a thought: Do we want to apply the `DamageTypes` here as well?

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Makes sense, yeah. Updated.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 4]
```diff
@@ -70,7 +74,7 @@ public void AddedToWorld(Actor self)
 			// Actors start with maximum damage applied
 			var delta = health.HP - damageThreshold;
 			if (delta > 0)
-				health.InflictDamage(self, self.World.WorldActor, new Damage(delta), false);
+				self.InflictDamage(self, new Damage(delta, Info.DamageTypes));
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: This isn't a suicide so I think you should revert this back to:

```csharp
self.InflictDamage(self.World.WorldActor, new Damage(Info.Damage, Info.DamageTypes));
```

I checked. It doesn't have a negative impact. Still fixes https://github.com/OpenRA/OpenRA/issues/11383.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 5]
```diff
@@ -86,8 +90,8 @@ public void Tick(Actor self)
 			if (!Info.Terrain.Contains(t.Type))
 				return;
 
-			Info.WeaponInfo.Impact(Target.FromActor(self), self.World.WorldActor, Enumerable.Empty<int>());
-			damageTicks = Info.WeaponInfo.ReloadDelay;
+			self.InflictDamage(self, new Damage(Info.Damage, Info.DamageTypes));
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: Same here. Please change attacker to `self.World.WorldActor`.

**Comment by Mailaender** [comment]
I can confirm this fixes https://github.com/OpenRA/OpenRA/issues/11383. :white_check_mark: However the balancing is completely off. Infantry now get insta-killed when touching a Tiberium covered cell in the TD mod.

**Comment by GraionDilach** [comment]
Oh god, when I resolved https://github.com/OpenRA/OpenRA/pull/11634#discussion_r71286413 I forgot to update the YAML tags to reflect the new name... I'm stupid.

File: OpenRA.Mods.Common/Traits/DamagedByTerrain.cs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-﻿#region Copyright & License Information
+#region Copyright & License Information
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/DamagedByTerrain.cs]: This removes a BOM - automated change by VS although from what I see, that's how it should be anyway.

**Comment by GraionDilach** [comment]
Fixed all issues aforementioned.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/287ff996a24f5ba28133f2bf692a0dd2386c476c...12a7a48f6b89bf8916a6260719a80b48bde4bfb6)

**Comment by Mailaender** [comment]
Thanks for taking care about this.

**Pull request merged**